### ISSUE TITLE
Add RSS feed to allow visitors to follow the project gallery

### ISSUE THREAD

**Issue by Vincent14** [sleede/fab-manager#23]
Hello, I'm a Ruby on Rails developer and user at Caen, in the Relais d'Science fablab.

I would like to be alerted for each new post in the project gallery. Last week I've wrote a tutorial to add RSS feed in a RoR action (REALLY easy) in only three steps: http://www.sois-net.fr/les-3-etapes-pour-creer-un-flux-rss-en-rails/

I've no idea how to contribute, and I'm not familiar with your frontend framework. However, if someone can guide me I'm ready to implement my request. If you've read my tutorial above and think it's doable quickly, I'll be very pleased :) 

This feature could also allow us to publish on social networks, until connexions would be available.

**Comment by nflorentin** [comment]
Hi and thanks you for your interest in the project!

That's a great idea. I invite you to read the document [CONTRIBUTING](https://github.com/LaCasemate/fab-manager/blob/master/CONTRIBUTING.md). 
After that, you can implement your feature and submit your pull request. 
I'd suggest to implement it in a new controller projects_controller.rb in app/controllers directory.

For the frontend part, no problem, you can implement the endpoint (ctrl+view) and we'll take care of adding the link on the frontend if you are struggling with it.

We can't wait to merge your contribution ! :)

**Comment by Vincent14** [comment]
Okai, I've installed fabmanager on Ubuntu and tried to make some changes. I followed my own tutorial (given in my first post) to add a proof of concept of RSS feed:

Here is my `app/views/api/projects/index.rss.builder` file:

```
#encoding: UTF-8

xml.instruct! :xml, version: "1.0"
xml.rss version: "2.0" do
  xml.channel do
    xml.title "Flux RSS de telle partie du site"
    xml.author "Nom du fablab"
    xml.description "Ce flux RSS liste les entrées de telle partie du site"
    xml.link "url de la page html"
    xml.language "fr"

    @projects.each do |project|
      xml.item do
        xml.guid project.id
        xml.pubDate project.created_at.strftime("%F %T")
        xml.title project.name
        xml.link project.url
        xml.author project.author.first_name
        xml.description project.description
      end
    end
  end
end
```

Saddly, the url supposed to return XML is actually returning JSON:

> http://www.localhost:3000/api/projects/?format=rss

`{"projects":[{"id":5,"name":"testestest","description":"<p>testestestes<br></p>","author_id":1,"licence_id":null,"slug":"testestest","url":"http://www.localhost:3000/api/projects/testestest","user_ids":[]},{"id":4,"name":"Again again","description":"<p>again desct<br></p>","author_id":1,"licence_id":null,"slug":"again-again","url":"http://www.localhost:3000/api/projects/again-again","user_ids":[]},{"id":3,"name":"test","description":"<p>descr<br></p>","author_id":1,"licence_id":null,"slug":"test","url":"http://www.localhost:3000/api/projects/test","project_image":"/uploads/project_image/3/medium_project_image.png","user_ids":[]},{"id":2,"name":"autre projet","description":"<p>autre description<br></p>","author_id":1,"licence_id":null,"slug":"autre-projet","url":"http://www.localhost:3000/api/projects/autre-projet","project_image":"/uploads/project_image/2/medium_project_image.png","user_ids":[]}]}`

Do you have any idea?

**Comment by nflorentin** [comment]
Hello,

did you try /api/projects.rss ?

**Comment by Vincent14** [comment]
I have a strange error (it's supposed to work out of the box):

```

NameError in API::Projects#index

Showing /media/Stockage/Projets/fab-manager/fab-manager/app/views/api/projects/index.rss.jbuilder where line #3 raised:

undefined local variable or method `xml' for #<#<Class:0x0000000848ccd8>:0x00000007d69d20>

Extracted source (around line #3):


#encoding: UTF-8

xml.instruct! :xml, version: "1.0"
xml.rss version: "2.0" do
  xml.channel do
    xml.title "Flux RSS de telle partie du site"

Rails.root: /media/Stockage/Projets/fab-manager/fab-manage
```

**Comment by sylvainbx** [comment]
[Jbuilder](https://github.com/rails/jbuilder) is only intended to generate JSON, not XML. Are you sure you're using the right gem or there's no collision ?

**Comment by Vincent14** [comment]
I did the tutorial on a regular project and hadn't anything to do in addition. But it wasn't a fresh new project, I'm just sure that no one of my colleagues know what is RSS, and no one did something in favor of this configuration.

It seems to be provided by an ActionView class helper (itself!) according to [this railcast](http://railscasts.com/episodes/87-generating-rss-feeds) and [this official doc page](http://apidock.com/rails/ActionView/Helpers/AtomFeedHelper/atom_feed)

**[breaking news]**: I found the problem! I named my file ".jbuilder" instead of ".builder"!

![capture du 2016-09-10 10-15-36](https://cloud.githubusercontent.com/assets/1963080/18409035/9c6c9996-773f-11e6-93bb-72b1305ed033.png)

**Comment by Vincent14** [comment]
Okai, let's see if I follow correctly your code design!

```
#encoding: UTF-8

xml.instruct! :xml, version: "1.0"
xml.rss version: "2.0" do
  xml.channel do
    xml.title "#{t('app.public.projects_list.the_fablab_projects')} - #{Setting.find_by(name: 'fablab_name').value}"
    xml.description t('app.public.projects_list.all_projects')
    xml.author Setting.find_by(name: 'fablab_name').value
    xml.link request.base_url + "/#!/projects"
    xml.language I18n.locale.to_s

    @projects.each do |project|
      xml.item do
        xml.guid project.id
        xml.pubDate project.created_at.strftime("%F %T")
        xml.title project.name
        xml.link request.base_url + "/#!/projects/" + project.slug
        xml.author project.author.first_name
        xml.description project.description
      end
    end
  end
end
```

Generates:

```
<?xml version="1.0" encoding="UTF-8"?>
<rss version="2.0">
  <channel>
    <title>Les projets du FabLab - Fab Lab de La Casemate</title>
    <description>Tous les projets</description>
    <author>Fab Lab de La Casemate</author>
    <link>http://www.localhost:3000/#!/projects</link>
    <language>fr</language>
    <item>
      <guid>5</guid>
      <pubDate>2016-09-03 13:05:29</pubDate>
      <title>testestest</title>
      <link>http://www.localhost:3000/#!/projects/testestest</link>
      <author>admin</author>
      <description>&lt;p&gt;testestestes&lt;br&gt;&lt;/p&gt;</description>
    </item>
    <item>
      <guid>4</guid>
      <pubDate>2016-09-01 18:29:18</pubDate>
      <title>Again again</title>
      <link>http://www.localhost:3000/#!/projects/again-again</link>
      <author>admin</author>
      <description>&lt;p&gt;again desct&lt;br&gt;&lt;/p&gt;</description>
    </item>
    <item>
      <guid>3</guid>
      <pubDate>2016-08-31 21:30:42</pubDate>
      <title>test</title>
      <link>http://www.localhost:3000/#!/projects/test</link>
      <author>admin</author>
      <description>&lt;p&gt;descr&lt;br&gt;&lt;/p&gt;</description>
    </item>
    <item>
      <guid>2</guid>
      <pubDate>2016-08-31 21:29:47</pubDate>
      <title>autre projet</title>
      <link>http://www.localhost:3000/#!/projects/autre-projet</link>
      <author>admin</author>
      <description>&lt;p&gt;autre description&lt;br&gt;&lt;/p&gt;</description>
    </item>
  </channel>
</rss>
```

I draw your attention on the way I retrieve fronted URL from the backend, and translated strings, is it ok for you?

**Comment by nflorentin** [comment]
Hi,

sorry for the delay, It'd be better to use url helpers instead of generating the url manually. There are helpers named `projects_url` and `project_url(@project)`. You can see the full list of routes/url helpers with the command `rake routes`.

**Comment by Vincent14** [comment]
I would add the link tag on the frontend, to make RSS feed available, where should I put this?
`<link rel="alternate" type="application/rss+xml" title="Flux RSS" href="/controller/action?format=rss">`

**Comment by Vincent14** [comment]
Basically I can't use `projects_url` and `project_url(@project)` because they generate the server side URL (the API one). The links I need are the frontend ones, because if RSS is generated from backend, links of content reachable by people are the ones of the frontend.

Pull request over here https://github.com/LaCasemate/fab-manager/pull/44 I've clone from `master` and suggest to push in `dev` branch, but a file I didn't change is now in the pull request ([ISSUE_TEMPLATE.md](https://github.com/LaCasemate/fab-manager/pull/44/files))

**Comment by sylvainbx** [comment]
Hi @Vincent14 
I've just merged your pull request into the dev branch (see https://github.com/LaCasemate/fab-manager/commit/8dbbbfad9edff969b95841880f4beca052fd1596). Congratz :+1: 

Before merging I've made two important changes:
- Moved your XML views from the `API` folder to a new `RSS` folder. As the api is json-oriented and this allows to use separate controllers, enhancing code sustainability
- Added links to your feeds into `app/views/application/index.html.erb`

Feel free to comment or open a new pull request if you'd like to improve this in the future.
This (https://github.com/LaCasemate/fab-manager/commit/34ad1c905604368425cedc5c81fbe4dabfeb0f38) may also gave you ideas about implementing machines and trainings availabilities feeds.
